README.md
# api-console-builder

__This version of the builder only works with API console version 6 and higher.__

[![Build Status](https://travis-ci.org/mulesoft-labs/api-console-builder.svg?branch=master)](https://travis-ci.org/mulesoft-labs/api-console-builder)

__api-console-builder__ allows to generate production ready bundle of MuleSoft [API Console](https://api-console.io).

On a high level, it allows to bundle a stand-alone application of API Console. If you need to embed API Console in your application follow instructions from [Rollup configuration for API Console](https://docs.api-console.io/building/rollup/).

## Usage

The bundler creates a production ready API Console that works in modern web browsers and in IE11 withj polyfilled support for web components.

```javascript
import { ApiConsoleProject } from '@api-components/api-console-builder';
const project = new ApiConsoleProject({
  destination: 'api-console-build',
  api: 'path/to/api.raml',
  apiType: 'RAML 1.0',
  apiMediaType: 'application/raml',
  tagName: '6.0.0'
});
await project.bundle();
```

## Supported API types

API console accepts AMF `json/ld` model as a data source. AMF by default supports the following API formats:

-   RAML 0.8
-   RAML 1.0
-   OAS 2.0
-   OAS 3.0 - Experimental, may be missing some properties.

## Configuration

The configuration is defined in [BuilderOptions.js](lib/BuilderOptions.js) file.

### tagName

Type: `string`

A release tag name to use. With this option the builder uses specific release of the console. If not set and `src` is not set it uses latest release. Note, only versions >= 6.0.0 can be used with this version of the builder.

### destination

Type: `string`

Output directory. Default to `build`.

### api

Type: `string`

__required__

Unlike previous versions of this library, now `api` property is required. The build will fail if this is not set.

Location of API specification main file.

Set `apiType` property to corresponding value (type of the API, see below).

Default to `undefined`.

### apiType

Type: `string`

_required__

Type of an API spec file recognizable by [AMF](https://github.com/mulesoft/amf).

### apiMediaType

Type: `string`

Media type of the API.

For RAML files it is always `application/yaml`. OAS comes with two flavours: `application/yaml` and `application/json`.

Use it when the library can't process API spec file due to processing error.

### verbose

Type: `boolean`

Prints a debug messages.

### attributes

Type: `array`

An array of attributes to be set on the `<api-console>` element.

For boolean attributes just add name of the attribute as string.

For attributes with values add a map where the key is the attribute name
and value is the attribute value.

Note: Do not use camel case notation. It will not work. See the example.

#### Example

```javascript
const attributes = [
 'proxyencodeurl',
 {'proxy': 'https://proxy.domain.com'},
 'notryit',
 {'page': 'request'},
]
```

Example above is the same as:

```javascript
const attributes = [
 'proxyencodeurl',
 'notryit',
 {
   'proxy': 'https://proxy.domain.com',
   'page': 'request'
 }
]
```

and will produce the following output:

```html
<api-console-app
 proxyencodeurl
 notryit
 page="request"
 proxy="https://proxy.domain.com"
></api-console-app>
```

List of all available options can be found here: https://github.com/mulesoft/api-console/blob/master/docs/configuring-api-console.md

Note, you don't need to set this property when providing own `indexFile`. Simply define attributes in the file.


### logger

Type: `object`

A console like object to print debug output. If not set then it creates it's own logger.

### themeFile

Type: `string`

Location to a theme file with styles definition of the console.
It replaces Console's own styles definition. See theming documentation of the API console for more information.

### indexFile

Type: `string`

Location to a custom `index.html` file that will be used instead of the default template.

The template file must include vendor package, API Console sources, and the use of API Console. See `templates/index.html` for an example.

### noCache

Type: `boolean`

By default the builder caches build results in user home folder and uses generated sources to speed up the build process.

Note, options that influence the build process (`tagName`, `themeFile`, `indexFile`, etc) creates new cached file.

API model is never cached.

### exitOnError

Type: `boolean`

Decides whether to finish current process with non-zero exit code on error. Default is `true`.

### strict

Type: `boolean`

When set to `true` it will stop build process with error when a minor issue has been detected, like missing theme file in declared location which is normally ignored.

### appTitle

Type: `string`

Optional application title put into HTML's `<title>` tag. By default it uses API title or `API Console` if missing.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.5](https://github.com/mulesoft-labs/api-console-builder/compare/v0.4.4...v2.0.5) (2020-01-31)


### Bug Fixes

* Adding fix for styles to the import file. ([25a4f50](https://github.com/mulesoft-labs/api-console-builder/commit/25a4f50ac91151abdc958d6a9d5e0f1cbfda50c5))
* Adding missing support for noOauth and OAuth element ([26dc181](https://github.com/mulesoft-labs/api-console-builder/commit/26dc181a1e7c6bbb4c488c2031ea91aba35c6b32))
* fixes bundling command for Windows platform ([63b7ae8](https://github.com/mulesoft-labs/api-console-builder/commit/63b7ae8e53003396e1cc166af5fef1f03b6ec80f))
* Fixes https://github.com/mulesoft-labs/api-console-cli/issues/29 ([deb3cc6](https://github.com/mulesoft-labs/api-console-builder/commit/deb3cc60c6c8c87651438b230e3c9ad4fbc3e866))
* fixing logger call to undefined function ([71d7cbe](https://github.com/mulesoft-labs/api-console-builder/commit/71d7cbebe58a78e4f5e0226c620f995027434f1f))
* Fixing npm audit and dependency version ([ac03277](https://github.com/mulesoft-labs/api-console-builder/commit/ac0327714126eb46336fd23688faaaac9e7513b2))
* Fixing script error ([d569f04](https://github.com/mulesoft-labs/api-console-builder/commit/d569f04d98209d93a3902e1989cdb85f07c39c68))
* Fixing stable build ([ed81de6](https://github.com/mulesoft-labs/api-console-builder/commit/ed81de6c8dd062ab43d620ef3ec05b2c36954ebf))
* Fixing test command to increase memory limit ([4dabb93](https://github.com/mulesoft-labs/api-console-builder/commit/4dabb93700ad1aa3558c267ea6d89618905095ab))
* Refactoring misspelling ([da13f77](https://github.com/mulesoft-labs/api-console-builder/commit/da13f777719b1a76e972fb69739b86b85b2a16e2))
* Upgrading dependencies for security vulnerability ([40efb2f](https://github.com/mulesoft-labs/api-console-builder/commit/40efb2f5e2d5ea136ac537d7575e6d37f505aa3a))

<a name="1.0.11"></a>
## 1.0.11 (2020-01-13)


* Bumping version ([e01d4cf](https://github.com/mulesoft-labs/api-console-builder/commit/e01d4cf))

### chore

* chore: adding main bundle process flow ([b23b89c](https://github.com/mulesoft-labs/api-console-builder/commit/b23b89c))
* chore: finalizing builder development ([5e5981a](https://github.com/mulesoft-labs/api-console-builder/commit/5e5981a))
* chore: moving vendor generation to the tmp folder ([dac93fa](https://github.com/mulesoft-labs/api-console-builder/commit/dac93fa))
* chore: starting rewriting the component for APIC v6 ([3e02f50](https://github.com/mulesoft-labs/api-console-builder/commit/3e02f50))

### Fix

* Fix: Fixing test command to increase memory limit ([4dabb93](https://github.com/mulesoft-labs/api-console-builder/commit/4dabb93))
* Fix: Upgrading dependencies for security vulnerability ([40efb2f](https://github.com/mulesoft-labs/api-console-builder/commit/40efb2f))

### Update

* Update: Adding additional debug output ([7606de3](https://github.com/mulesoft-labs/api-console-builder/commit/7606de3))
* Update: Removing theme installation as it is installed with APIC ([e08ad02](https://github.com/mulesoft-labs/api-console-builder/commit/e08ad02))
* Update: Restoring theme installation ([890308e](https://github.com/mulesoft-labs/api-console-builder/commit/890308e))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/mulesoft-labs/api-console-builder/compare/0.3.0...v0.4.0) (2017-06-30)


### Breaking

* Bumped version to 0.4.0 as incompatible with previous release. ([4fbf718](https://github.com/mulesoft-labs/api-console-builder/commit/4fbf718))
* Renamed source control file. ([e7b6732](https://github.com/mulesoft-labs/api-console-builder/commit/e7b6732))
* Renamed the options class. ([b0d112c](https://github.com/mulesoft-labs/api-console-builder/commit/b0d112c))
* Replaced options that are related to setting an attributes on th ([463b1b2](https://github.com/mulesoft-labs/api-console-builder/commit/463b1b2))

### Docs

* Updated documentation for the module. ([4da4306](https://github.com/mulesoft-labs/api-console-builder/commit/4da4306))

### Fix

* Added missing argument to function all. ([68f4f55](https://github.com/mulesoft-labs/api-console-builder/commit/68f4f55))
* Added return statements to the function. ([44797fc](https://github.com/mulesoft-labs/api-console-builder/commit/44797fc))
* Fixed an issue on Ubuntu bash on windows which by default runs as roo ([c6b3de5](https://github.com/mulesoft-labs/api-console-builder/commit/c6b3de5))
* Fixed issue when reading redirection headers. ([b506248](https://github.com/mulesoft-labs/api-console-builder/commit/b506248))
* Fixed issue with used API. The new form of the URL class is too new.  ([9b9d997](https://github.com/mulesoft-labs/api-console-builder/commit/9b9d997))
* Fixes #1 - changed logic for getting the sources and processing them. ([28e4537](https://github.com/mulesoft-labs/api-console-builder/commit/28e4537)), closes [#1](https://github.com/mulesoft-labs/api-console-builder/issues/1)
* Reversed order when preparing build so the zip files are handled corr ([6407e89](https://github.com/mulesoft-labs/api-console-builder/commit/6407e89))

### New

* Added new `jsonFile` option to point a file name with RAML JSON data. ([4b1601e](https://github.com/mulesoft-labs/api-console-builder/commit/4b1601e))
* Added new class responsible for handling local templates and updating ([0cbe1c6](https://github.com/mulesoft-labs/api-console-builder/commit/0cbe1c6))
* Added new class responsible for setting attributes on the API console ([5b0bff1](https://github.com/mulesoft-labs/api-console-builder/commit/5b0bff1))
* added new classes to support querying and transport to GitHub. ([c0ecfa1](https://github.com/mulesoft-labs/api-console-builder/commit/c0ecfa1))

lib/amf-source.js
'use strict';
/**
 * Copyright (C) Mulesoft.
 * Shared under Apache 2.0 license
 *
 * @author Pawel Psztyc
 */
const amf = require('amf-client-js');
const fs = require('fs-extra');
const path = require('path');
/**
 * A class to generate AMF json/ld data model.
 * It is using the "editing" pipeline created for the console
 * by the AMF team.
 */
class AmfSource {
  /**
   * @param {Object} logger
   */
  constructor(logger) {
    this.logger = logger;
  }
  /**
   * Generates the json/ld model from an API file.
   * @param {String} location API file location
   * @param {String} type API type.
   * @return {Promise} Promise resolved to a model.
   */
  getModel(location, type) {
    this.logger.info(
      `Generating API model from ${location}, using ${type} parser`);
    amf.plugins.document.WebApi.register();
    amf.plugins.document.Vocabularies.register();
    amf.plugins.features.AMFValidation.register();

    this.logger.debug('Initializing AMF library...');
    return amf.Core.init()
    .then(() => {
      this.logger.debug('AMF ready.');
      this.logger.debug('Running API parser...');
      const parser = amf.Core.parser(type, 'application/yaml');
      let url;
      if (location.indexOf('http') === 0) {
        url = location;
      } else {
        url = `file://${location}`;
      }
      return parser.parseFileAsync(url);
    })
    .then((doc) => {
      this.logger.debug('API parsed.');
      this.logger.debug('Validating API...');
      let validateProfile;
      switch (type) {
        case 'RAML 1.0': validateProfile = amf.ProfileNames.RAML; break;
        case 'RAML 0.8': validateProfile = amf.ProfileNames.RAML08; break;
        case 'OAS 2.0':
        case 'OAS 3.0':
          validateProfile = amf.ProfileNames.OAS;
          break;
      }
      return amf.AMF.validate(doc, validateProfile)
      .then((report) => {
        if (!report.conforms) {
          this.logger.warn(report.toString());
        } else {
          this.logger.debug('API valid.');
        }
        return doc;
      });
    })
    .then((doc) => {
      this.logger.debug('Resolving API model for API components...');
      const resolver = amf.Core.resolver(type);
      return resolver.resolve(doc, 'editing');
    });
  }
  /**
   * Saves the model to a file.
   * @param {Array} model json/ld model
   * @param {String} file Path to a file where to savbe the model.
   * @return {Promise}
   */
  saveModel(model, file) {
    this.logger.debug('Generating json-ld model...');
    const opts = amf.render.RenderOptions().withSourceMaps.withCompactUris;
    const generator = amf.Core.generator('AMF Graph', 'application/ld+json');
    const start = Date.now();
    return generator.generateString(model, opts)
    .then((data) => {
      const time = Date.now() - start;
      this.logger.debug(`Model ready in ${time} milliseconds`);
      this.logger.debug('Storing API data model to file: ' + file);
      const dir = path.dirname(file);
      return fs.ensureDir(dir)
      .then(() => fs.writeFile(file, data, 'utf8'));
    });
  }
}

exports.AmfSource = AmfSource;
